File: src/cbridge/GSL.m
//
//  GSL.m
//  grease — GNU Scientific Library wrappers (implementations)
//

#include "GSL.h"
#include <gsl/gsl_sf_gamma.h>
#include <gsl/gsl_sf_erf.h>
#include <gsl/gsl_sf_bessel.h>
#include <gsl/gsl_sf_log.h>
#include <gsl/gsl_sf_exp.h>
#include <gsl/gsl_sf_zeta.h>
#include <gsl/gsl_pow_int.h>

double grease_gsl_sf_gamma(double x)       { return gsl_sf_gamma(x); }
double grease_gsl_sf_erf(double x)         { return gsl_sf_erf(x); }
double grease_gsl_sf_bessel_J0(double x)   { return gsl_sf_bessel_J0(x); }
double grease_gsl_sf_bessel_J1(double x)   { return gsl_sf_bessel_J1(x); }
double grease_gsl_sf_log(double x)         { return gsl_sf_log(x); }
double grease_gsl_sf_exp(double x)         { return gsl_sf_exp(x); }
double grease_gsl_sf_zeta(double s)        { return gsl_sf_zeta(s); }
double grease_gsl_pow_int(double x, int n) { return gsl_pow_int(x, n); }

File: dev/calc_app/gsl.clj
(ns calc-app.gsl
  "Clojure wrappers around GSL C functions exposed via Bridge.m / GSL.m.
  Each function calls through `ffi/call` using dlsym(RTLD_DEFAULT, name)
  which finds the statically-linked GSL symbols at runtime."
  (:require [com.phronemophobic.clj-libffi :as ffi]))

(defn gamma
  "Gamma(x) — Γ(n) = (n-1)! for positive integers."
  [x]
  (ffi/call "grease_gsl_sf_gamma" :float64 :float64 x))

(defn erf
  "Error function erf(x) — integral of Gaussian, range [-1, 1]."
  [x]
  (ffi/call "grease_gsl_sf_erf" :float64 :float64 x))

(defn bessel-J0
  "Bessel function of the first kind, order 0: J₀(x)."
  [x]
  (ffi/call "grease_gsl_sf_bessel_J0" :float64 :float64 x))

(defn bessel-J1
  "Bessel function of the first kind, order 1: J₁(x)."
  [x]
  (ffi/call "grease_gsl_sf_bessel_J1" :float64 :float64 x))

(defn ln
  "Natural logarithm ln(x) via GSL."
  [x]
  (ffi/call "grease_gsl_sf_log" :float64 :float64 x))

(defn exp-gsl
  "Exponential e^x via GSL."
  [x]
  (ffi/call "grease_gsl_sf_exp" :float64 :float64 x))

(defn pow-int
  "x raised to integer power n (exact, no floating-point exponent)."
  [x n]
  (ffi/call "grease_gsl_pow_int" :float64 :float64 :sint32 x (int n)))

(defn zeta
  "Riemann zeta function ζ(s)."
  [s]
  (ffi/call "grease_gsl_sf_zeta" :float64 :float64 s))
